service/interpreter/stateExecutionCounter.go

```diff
@@ -180,6 +191,23 @@ func (e *StateExecutionCounter) updateStateIdSearchAttribute() error {
 		// see workflowImpl.go to call ClearExecutingStateIdsSearchAttributeFinally at the end
 		return nil
 	}
+
+	if stateTransition != nil {
+		// UpsertSearchAttributes should be only invoked if when the next states do not skip waitUntil
+		shouldSkipUpsertingSAs := true
+		for _, s := range stateTransition.next {
+			if !s.StateOptions.GetSkipWaitUntil() {
+				shouldSkipUpsertingSAs = false
```

Comment:
I think perhaps I'm missing some context or understanding of what we want to achieve here - what happens when the first StateMovement in the next array skips waitUntil, but subsequent ones do not?  If I'm reading this correctly, with this change if any next StateMovement does not have a waitUntil, the `SearchAttributeExecutingStateIds` will not be updated...

Reply:
hmm... I have that wrong don't I... if any of the next states do have a waitUntil (i.e. they don't skip wait until) then do continue on to UpsertSearchAttributes??

Reply:
Right, if any of the next states don't skip waitUntil, then the SAs will be upserted. Also, if I'm not mistaken, nextStates are the states that are executed in the state transition, example:
current = A
next = [B, C, D]
means that B, C, and D are executed at the same time after A is completed.